### PR TITLE
Add pure-Haskell replacement for is_ascii.c

### PR DESCRIPTION
**Pull request by chreekat**
wip. Will add some comments inline.

### CODE REVIEW

File: src/Data/Text/Encoding.hs [hunk 1]
```diff
@@ -219,6 +222,9 @@ decodeLatin1 ::
   HasCallStack =>
 #endif
   ByteString -> Text
+#if defined(PURE_HASKELL)
+decodeLatin1 bs = pack (Data.ByteString.Char8.unpack bs)
```
Comment (chreekat) [inline src/Data/Text/Encoding.hs:226]: Thanks to the Haddocks for suggesting this quick replacement :D

File: src/Data/Text/Internal/IsAscii.hs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+--
+-- >>> isAscii "abcd€"
+-- False
+--
+-- prop> isAscii t == all (< '\x80') t
+--
+-- @since 2.0.2
+isAscii :: Text -> Bool
+#if defined(PURE_HASKELL)
+isAscii = P.const False
```
Comment (chreekat) [inline src/Data/Text/Internal/IsAscii.hs:48]: todo

File: src/Data/Text/Internal/IsAscii.hs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+isAscii (Text (A.ByteArray arr) off len) =
+    cSizeToInt (c_is_ascii_offset arr (intToCSize off) (intToCSize len)) == len
+#endif
+{-# INLINE isAscii #-}
+
+-- | Length of the longest ASCII prefix.
+asciiPrefixLength :: ByteString -> Int
+#if defined(PURE_HASKELL)
+asciiPrefixLength = P.const 0
```
Comment (chreekat) [inline src/Data/Text/Internal/IsAscii.hs:65]: todo

File: text.cabal [hunk 4]
```diff
@@ -93,11 +93,17 @@ flag pure-haskell
   default: False
   manual: True
 
+flag repl-tests
```
Comment (chreekat) [inline text.cabal:96]: I added this so I could run the tests quickly with `ghcid`.

**Review by chreekat — COMMENTED** [review]
It looks like the impacted exported functions are `decodeLatin1` and `isAscii`.

**Comment by Bodigrim** [comment]
@chreekat @hsyl20 are we looking to finish this up in time for GHC 9.10 (which is soon)?

**Comment by chreekat** [comment]
@Bodigrim for me at least, this is just a free-time activity, and I haven't had a lot of that lately. I know @hsyl20 's team has been working an a pure-Haskell bytestring that they want to get finished for 9.10, but not sure about text.

**Comment by hsyl20** [comment]
Nothing started about `text` afaik. So it's likely that it won't be done in time for 9.10

File: text.cabal [hunk 5]
```diff
@@ -258,7 +265,7 @@ source-repository head
 test-suite tests
   type:           exitcode-stdio-1.0
   ghc-options:
-    -Wall -threaded -rtsopts
+    -Wall -threaded -rtsopts -with-rtsopts=-N
```
Comment (chreekat) [inline text.cabal:268]: Let me know if there's a good reason this wasn't already here. It's not related to the PR.

**Comment by chreekat** [comment]
Plot twist: I had some extra time today.

**Comment by chreekat** [comment]
Oops, bad rebase. Will fix in the morning.

**Comment by chreekat** [comment]
This is ready for review now.

File: src/Data/Text/Internal/IsAscii.hs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+--
+-- >>> isAscii "abcd€"
+-- False
+--
+-- prop> isAscii t == all (< '\x80') t
+--
+-- @since 2.0.2
+isAscii :: Text -> Bool
+#if defined(PURE_HASKELL)
+isAscii = all Char.isAscii
```
Comment (Bodigrim) [inline src/Data/Text/Internal/IsAscii.hs:51]: I'd rather avoid complications with `Data.Text.hs-boot` file. It's ugly, but one can write a loop via [`iter`](https://hackage.haskell.org/package/text-2.1/docs/Data-Text-Unsafe.html#v:iter), checking if each symbol `isAscii`.

Comment (chreekat) [reply src/Data/Text/Internal/IsAscii.hs:54]: done

File: text.cabal [hunk 7]
```diff
@@ -93,11 +93,17 @@ flag pure-haskell
   default: False
   manual: True
 
+flag repl-tests
+  description: Enable running tests in the repl
```
Comment (Bodigrim) [inline text.cabal:97]: @chreekat How do you invoke tests in the repl? Sorry, I have not used `ghcid` for ages and failed to figure out myself.

Comment (Bodigrim) [reply text.cabal:97]: Aha, I suppose it's something like `ghcid --command='cabal v2-repl test:tests' --test 'Main.main'`.

Comment (Bodigrim) [reply text.cabal:97]: I've just published `tasty-inspection-testing-0.2.1` which detects if it is executed inside ghci and skips inspections tests automatically. Could you please `cabal update` and check the flag is still needed?

Comment (chreekat) [reply text.cabal:97]: Yes, that did work, but to be honest I was also using the flag to run the tests with --disable-optimisation. :) 

That being the case, the flag has a bad name.

I'll just remove it from this PR so we can think about it more in a different context.

**Review by Bodigrim — APPROVED** [review]

**Review by Lysxia — APPROVED** [review]

**Comment by hsyl20** [comment]
Thanks for doing this @chreekat!

**Comment by Bodigrim** [comment]
Thanks!

**Pull request merged**